### ISSUE TITLE
E2E Test Suite & Chaos

### ISSUE THREAD

**Issue by shaneutt** [Kong/kubernetes-ingress-controller#1605]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

We have integration tests that test against local development clusters like `kind`  and have specific feature related test cases, but we don't have end-to-end (e2e) test cases that cover a longer lifecycle of components within the cluster, interactions between various components, and on conformant and production quality Kubernetes clusters.

### Proposed Solution

- [ ] implement a full E2E test suite using GKE
- [ ] integrate the test suite into our release workflow as a release blocker

### Additional information

Our e2e tests should try to mimic a production lifecycle of components within the cluster, the controller, and the proxy. We should introduce chaos to simulate production cluster problems.

Specific Factors:
- [ ] tests of our APIs in the e2e environment should be pointed at real applications (e.g. not `httpbin`) doing validated work
- [ ] the cluster should have its resources exhausted at various points, semi-arbitrarily
- [ ] the cluster network should be saturated at various points
- [ ] nodes should be randomly killed without warning during the tests

### Acceptance Criteria

- [ ] as a contributor I have confidence that my changes will be tested with simulated production workloads (and failures) before release

**Comment by shaneutt** [comment]
We've now got E2E tests. Chaos is something we may revisit at a later time.

**Comment by randmonkey** [comment]
reopen this since we have e2e tests now.

**Comment by pmalek** [comment]
Perhaps we missed the "chaos testing" aspect when discussing this issue today.

The idea was to handle the e2e aspect, namely addressing: https://github.com/Kong/kubernetes-ingress-controller/blob/4bb94699c5c28a632c5ea1201c9b951e38a1d209/hack/e2e/run-tests.sh#L3-L7

, our take on which was that this can basically be removed since [we have e2e tests](https://github.com/Kong/kubernetes-ingress-controller/tree/main/test/e2e) now, so we don't need this "workaround" by running integration tests in place of e2e ones.

If you feel this is an incorrect approach please let us know @shaneutt

**Comment by shaneutt** [comment]
No that seems fine, but maybe just create a separate issue for that and keep this one closed since this comes with a lot of additional context and weight.

**Comment by mflendrich** [comment]
Some unpacking needed here.

This issue was discussed (together with #2751) in the team call today. There are following open problems:
1. relative mess between the below 3 classes of tests:
    - "integration tests" which are essentially blackbox tests of KIC within KIND
    - "e2e tests" being a script wrapper around the above integration tests that allow running in GKE
    - "e2e tests" being tests that test the integration between KIC and {Istio|Mesh|etc.}.
2. the open FR for "chaos monkey"-style testing (which I believe this issue was primarily about)
3. inconsistent mapping between test classes (listed in point 1 above) and CI workflows (PR gate, nightly, release).

The reason this was opened was because @randmonkey brought up the need to address (1) and (3).
Our recommended plan is to rewrite 2 issues (this one as well as #2751) into ones that would accurately capture the problems listed above. Hence the reopening and assignment. This issue is in "Awaiting AC & Prioritization" meaning that I intend to actively work on rewriting those. This would lead to the scope of (1) and (3) being written up with AC and essentially ready for pickup when we feel it's the time, and (2) being closed as a nice to have.

In other words: I'd like to carefully carve out isolated subproblems out of those two broad issues and be very tactical about which testing problems we solve, so that these problems are small and manageable.

**Comment by mflendrich** [comment]
> just create a separate issue for that and keep this one closed since this comes with a lot of additional context and weight

That's exactly the plan. :smile_cat:

**Comment by mflendrich** [comment]
To be handled in the following two issues (let me know if any part of the scope is missing):

The chaos part rewritten for narrower scope: #2787
The E2E test naming issue rewritten for narrower scope: #2788
